File: src/main/java/dev/psygamer/wireframe/util/TagCompound.java
package dev.psygamer.wireframe.util;

import com.google.common.primitives.Bytes;
import com.google.common.primitives.Ints;
import com.google.common.primitives.Longs;
import net.minecraft.nbt.CompoundNBT;

import java.util.Collection;
import java.util.UUID;

public class TagCompound {
	
	private final net.minecraft.nbt.CompoundNBT internal;
	
	public TagCompound() {
		this.internal = new net.minecraft.nbt.CompoundNBT();
	}
	
	private TagCompound(final net.minecraft.nbt.CompoundNBT internal) {
		this.internal = internal;
	}
	
	public static TagCompound get(final net.minecraft.nbt.CompoundNBT internal) {
		return new TagCompound(internal);
	}
	
	public void putByte(final String key, final byte value) {
		this.internal.putByte(key, value);
	}
	
	public byte getByte(final String key) {
		return this.internal.getByte(key);
	}
	
	public void putByteArray(final String key, final byte... value) {
		this.internal.putByteArray(key, value);
	}
	
	public void putByteArray(final String key, final Collection<Byte> value) {
		this.internal.putByteArray(key, Bytes.toArray(value));
	}
	
	public byte[] getByteArray(final String key) {
		return this.internal.getByteArray(key);
	}
	
	public void putShort(final String key, final short value) {
		this.internal.putShort(key, value);
	}
	
	public short getShort(final String key) {
		return this.internal.getShort(key);
	}
	
	public void putInt(final String key, final int value) {
		this.internal.putInt(key, value);
	}
	
	public int getInt(final String key) {
		return this.internal.getInt(key);
	}
	
	public void putIntArray(final String key, final int... value) {
		this.internal.putIntArray(key, value);
	}
	
	public void putIntArray(final String key, final Collection<Integer> value) {
		this.internal.putIntArray(key, Ints.toArray(value));
	}
	
	public int[] getIntArray(final String key) {
		return this.internal.getIntArray(key);
	}
	
	public void putLong(final String key, final long value) {
		this.internal.putLong(key, value);
	}
	
	public long getLong(final String key) {
		return this.internal.getLong(key);
	}
	
	public void putLongArray(final String key, final long... value) {
		this.internal.putLongArray(key, value);
	}
	
	public void putLongArray(final String key, final Collection<Long> value) {
		this.internal.putLongArray(key, Longs.toArray(value));
	}
	
	public long[] getLongArray(final String key) {
		return this.internal.getLongArray(key);
	}
	
	public void putFloat(final String key, final float value) {
		this.internal.putFloat(key, value);
	}
	
	public float getFloat(final String key) {
		return this.internal.getFloat(key);
	}
	
	public void putDouble(final String key, final double value) {
		this.internal.putDouble(key, value);
	}
	
	public double getDouble(final String key) {
		return this.internal.getDouble(key);
	}
	
	public void putBoolean(final String key, final boolean value) {
		this.internal.putBoolean(key, value);
	}
	
	public boolean getBoolean(final String key) {
		return this.internal.getBoolean(key);
	}
	
	public void putString(final String key, final String value) {
		this.internal.putString(key, value);
	}
	
	public String getString(final String key) {
		return this.internal.getString(key);
	}
	
	public void putUUID(final String key, final UUID value) {
		this.internal.putUUID(key, value);
	}
	
	public UUID getUUID(final String key) {
		return this.internal.getUUID(key);
	}
	
	public boolean contains(final String key) {
		return this.internal.contains(key);
	}
	
	public void remove(final String key) {
		this.internal.remove(key);
	}
	
	public boolean isEmpty() {
		return this.internal.isEmpty();
	}
	
	public CompoundNBT getInternal() {
		return this.internal;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/TagCompound.kt
package dev.psygamer.wireframe.util

import com.google.common.primitives.*
import java.util.*
import dev.psygamer.wireframe.nativeapi.wfWrapped

class TagCompound(
	internal val mcNative: net.minecraft.nbt.CompoundNBT = net.minecraft.nbt.CompoundNBT(),
) {

	val allKeys: Set<String> = this.mcNative.allKeys
	val isEmpty: Boolean = this.mcNative.isEmpty

	fun setByte(key: String, value: Byte) {
		this.mcNative.putByte(key, value)
	}

	fun getByte(key: String): Byte {
		return this.mcNative.getByte(key)
	}

	fun setByteArray(key: String, vararg value: Byte) {
		this.mcNative.putByteArray(key, value)
	}

	fun setByteArray(key: String, value: Collection<Byte>) {
		this.mcNative.putByteArray(key, Bytes.toArray(value))
	}

	fun getByteArray(key: String): ByteArray {
		return this.mcNative.getByteArray(key)
	}

	fun setShort(key: String, value: Short) {
		this.mcNative.putShort(key, value)
	}

	fun getShort(key: String): Short {
		return this.mcNative.getShort(key)
	}

	fun setInt(key: String, value: Int) {
		this.mcNative.putInt(key, value)
	}

	fun getInt(key: String): Int {
		return this.mcNative.getInt(key)
	}

	fun setIntArray(key: String, vararg value: Int) {
		this.mcNative.putIntArray(key, value)
	}

	fun setIntArray(key: String, value: Collection<Int>) {
		this.mcNative.putIntArray(key, Ints.toArray(value))
	}

	fun getIntArray(key: String): IntArray {
		return this.mcNative.getIntArray(key)
	}

	fun setLong(key: String, value: Long) {
		this.mcNative.putLong(key, value)
	}

	fun getLong(key: String): Long {
		return this.mcNative.getLong(key)
	}

	fun setLongArray(key: String, vararg value: Long) {
		this.mcNative.putLongArray(key, value)
	}

	fun setLongArray(key: String, value: Collection<Long>) {
		this.mcNative.putLongArray(key, Longs.toArray(value))
	}

	fun getLongArray(key: String): LongArray {
		return this.mcNative.getLongArray(key)
	}

	fun setFloat(key: String, value: Float) {
		this.mcNative.putFloat(key, value)
	}

	fun getFloat(key: String): Float {
		return this.mcNative.getFloat(key)
	}

	fun setDouble(key: String, value: Double) {
		this.mcNative.putDouble(key, value)
	}

	fun getDouble(key: String): Double {
		return this.mcNative.getDouble(key)
	}

	fun setBoolean(key: String, value: Boolean) {
		this.mcNative.putBoolean(key, value)
	}

	fun getBoolean(key: String): Boolean {
		return this.mcNative.getBoolean(key)
	}

	fun setString(key: String, value: String) {
		this.mcNative.putString(key, value)
	}

	fun getString(key: String): String {
		return this.mcNative.getString(key)
	}

	fun setUUID(key: String, value: UUID) {
		this.mcNative.putUUID(key, value)
	}

	fun getUUID(key: String): UUID {
		return this.mcNative.getUUID(key)
	}

	fun setCompound(key: String, compound: TagCompound) {
		this.mcNative.put(key, compound.mcNative)
	}

	fun getCompound(key: String): TagCompound {
		return this.mcNative.getCompound(key).wfWrapped
	}

	fun remove(key: String) {
		this.mcNative.remove(key)
	}

	operator fun contains(key: String): Boolean {
		return this.mcNative.contains(key)
	}
}
